Extension/src/LanguageServer/copilotCompletionContextProvider.ts

```diff
@@ -0,0 +1,309 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+import { CodeSnippet, ContextResolver, ResolveRequest } from '@github/copilot-language-server';
+import { randomUUID } from 'crypto';
+import * as vscode from 'vscode';
+import { DocumentSelector } from 'vscode-languageserver-protocol';
+import { getOutputChannelLogger, Logger } from '../logger';
+import * as telemetry from '../telemetry';
+import { CopilotCompletionContextResult } from './client';
+import { CopilotCompletionContextTelemetry } from './copilotCompletionContextTelemetry';
+import { getCopilotApi } from './copilotProviders';
+import { clients } from './extension';
+import { CppSettings } from './settings';
+
+class DefaultValueFallback extends Error {
+    static readonly DefaultValue = "DefaultValue";
+    constructor() { super(DefaultValueFallback.DefaultValue); }
+}
+
+class CancellationError extends Error {
+    static readonly Canceled = "Canceled";
+    constructor() { super(CancellationError.Canceled); }
+}
+
+class CopilotContextProviderException extends Error {
+}
+
+class WellKnownErrors extends Error {
+    static readonly ClientNotFound = "ClientNotFound";
+    private constructor(message: string) { super(message); }
+    public static clientNotFound(): Error {
+        return new WellKnownErrors(WellKnownErrors.ClientNotFound);
+    }
+}
+
+export enum CopilotCompletionContextFeatureFlag {
+    None = 0,
+    Instant,
+    Deferred
+}
+
+// Mutually exclusive values for the kind of returned completion context. They either are:
+// - computed.
+// - obtained from the cache.
+// - available in cache but stale (e.g. actual context is far away).
+// - missing since the computation took too long and no cache is present (cache miss). The value
+//   is asynchronously computed and stored in cache.
+// - the token is signaled as cancelled, in which case all the operations are aborted.
+// - an unknown state.
+export enum CopilotCompletionKind {
+    Computed = 'computed',
+    GotFromCache = 'gotFromCacheHit',
+    StaleCacheHit = 'staleCacheHit',
+    MissingCacheMiss = 'missingCacheMiss',
+    Canceled = 'canceled',
+    Unknown = 'unknown'
+}
+
+type CacheEntry = [string, CopilotCompletionContextResult];
+
+export class CopilotCompletionContextProvider implements ContextResolver<CodeSnippet> {
+    private static readonly providerId = 'cppTools';

```

Comment:
Where did you get this from? Is this supposed to be the extensionId? Our extension id is "ms-vscode.cpptools" and not "cppTools".

Reply:
@sean-mcmanus correct, it is certainly a good idea to use the extension id as a prefix. Since we may have more than one handler, we may have a custom suffix too, ideas? @benmcmorran @kuchungmsft @spebl